### PR TITLE
Allow chainable fixture adding in TestCase

### PR DESCRIPTION
**Pull request by dereuromark**
RFC
What do you think about allowing chainable adding of fixtures in TestCases using `->addFixture()` instead of the array property directly?
This would also allow autocompleting on it then in IDEs.

           /**
             * @return void
             */
            public function initialize(): void
            {
                $this->addFixture('core.Articles')
                    ->addFixture('app.SomeModel')
                    ->addFixture('plugin.Some/ComplicatedName.SomePluginModel');
            }

`initialize()` is called from the constructor as it has to be quite early for us allow setting them as such, otherwise we get a "table not found" error it seems.

We cannot use `setUp()` as it is too late.

### CODE REVIEW

File: src/TestSuite/TestCase.php [hunk 1]
```diff
@@ -80,6 +80,25 @@ abstract class TestCase extends BaseTestCase
...
     protected $_configure = [];
 
+    /**
+     * @param string $name Name
+     * @param array $data Data
+     * @param string $dataName Data name
+     */
+    public function __construct($name = null, array $data = [], $dataName = '')
+    {
+        $this->initialize();
```
Comment (markstory) [inline src/TestSuite/TestCase.php:90]: Could we call this method from `getFixtures()` instead?  That method was added in 4.0 to enable defining fixtures outside of the property.

Comment (dereuromark) [reply src/TestSuite/TestCase.php:90]: My main concern with the getFixtures() method was the side effect of having to call add..() inside.
But I guess that's a fair compromise for now.

**Comment by dereuromark** [comment]
I changed it so you can now extend `getFixtures()` directly:

       /**
         * @return string[]
         */
        public function getFixtures(): array
        {
            $this->addFixture('core.Articles')
                ->addFixture('app.SomeModel')
                ->addFixture('plugin.Some/ComplicatedName.SomePluginModel');

            return parent::getFixtures();
        }

etc

File: src/TestSuite/TestCase.php [hunk 2]
```diff
@@ -898,6 +898,24 @@ public static function setAppNamespace(string $appNamespace = 'TestApp'): ?strin
...
+     *
+     * Examples:
+     * - core.Tags
+     * - app.MyRecords
+     * - plugin.MyPluginName.MyModelName
+     *
+     * @param string $fixture Fixture
+     * @return $this
+     */
+    public function addFixture(string $fixture)
```
Comment (dereuromark) [inline src/TestSuite/TestCase.php:912]: Should this be protected? Any reason to have this public?

Comment (ADmad) [reply src/TestSuite/TestCase.php:912]: Can't think of any reason for it to be public.

Comment (dereuromark) [reply src/TestSuite/TestCase.php]: Fixed.

**Comment by dereuromark** [comment]
We can either backport this to 3.next - or use shimming ( https://github.com/dereuromark/cakephp-shim/pull/76 ) as opt-in.

File: src/TestSuite/TestCase.php [hunk 3]
```diff
@@ -898,6 +898,24 @@ public static function setAppNamespace(string $appNamespace = 'TestApp'): ?strin
         return $previous;
     }
 
+    /**
+     * Adds a fixture to this test case.
+     *
+     * Examples:
+     * - core.Tags
+     * - app.MyRecords
+     * - plugin.MyPluginName.MyModelName
```
Comment (markstory) [inline src/TestSuite/TestCase.php:907]: Should we give some guidance on how this method is to be used? For example 'Use this method inside your test cases' `getFixtures()` method to build up the fixture list'

Comment (dereuromark) [reply src/TestSuite/TestCase.php:907]: Sounds good. Here on in docs?

**Comment by ADmad** [comment]
If one wants IDE autocompletion can't they just use FQCNs?

**Comment by dereuromark** [comment]
Nah, thats the opposite of autocomplete, as you just get thousands of classes instead of the magic strings.
Plus a ton of use statement mess on top of it.

File: src/TestSuite/TestCase.php (hunk 3 above)
Comment (dereuromark) [reply src/TestSuite/TestCase.php:907]: I added the note in the docblock.

**Comment by chinpei215** [comment]
I agree with @ADmad . If people don't want thousands of classes or a ton of use statements,  they can do like the following:

```php
use Cake\Test\Fixture as f;
```
```php
public $fixtures = [
    f\ArticlesFixture::class,
    f\CommentsFixture::class,
];
```

**Comment by dereuromark** [comment]
That is the worst idea. It obfuscates everything.
Also, see the complex use cases as outlined in https://github.com/dereuromark/cakephp-ide-helper/pull/203 (nested ones)
To find them manually is way more difficult than using the documented and de facto standard string usage.

**Comment by markstory** [comment]
>  The one method doesn't cost anything, and is fully opt-in to be used (or not).

The one potential cost is people trying to use it inside a setup() or test method and it not working. In addition to not working it doesn't fail either adding a silent 'failure'.

**Comment by dereuromark** [comment]
Thus the documention added as per your suggestion.

**Comment by markstory** [comment]
I agree that using imports would be simpler for us, but having that many imports in a file could get unwieldy. This seems like a low-maintenance method for us to provide and help those who use IDE autocompletion.

**Pull request merged**